lib/spack/spack/database.py

```diff
@@ -339,7 +365,141 @@ def read_transaction(self):
+        if not spec.concrete:
+            raise ValueError('Concrete spec required for failure path for {0}'
+                             .format(spec.name))
+
+        return os.path.join(self._failure_dir,
+                            '{0}-{1}'.format(spec.name, spec.full_hash()))
+
+    def clear_failure(self, spec, force=False):
+        """Remove any persistent and cached failure tracking for the spec."""
+        if self.prefix_failure_locked(spec):
+            if not force:
+                tty.log('Retaining failure marking for {0} due to lock'
+                        .format(spec.name))
+                return
+            else:
+                tty.warn('Removing failure marking despite lock for {0}'
+                         .format(spec.name))
+
+        if spec.prefix in self._prefix_failures:
+            del self._prefix_failures[spec.prefix]
+
+        if self.prefix_failure_marked(spec):
+            try:
+                path = self._failed_spec_path(spec)
+                # TODO: Change to debug
+                tty.msg('Removing failure marking for {0}'.format(spec.name))
+                os.remove(path)
+            except Exception as exc:
+                tty.warn('Unable to remove failure marking for {0} ({1}): {2}'
+                         .format(spec.name, path, str(exc)))
+
+    def mark_failed(self, spec):
+        """
+        Mark a spec as failing to install.
+
+        Prefix failure marking takes the form of a byte range lock on the nth
+        byte of a file for coordinating between concurrent parallel build
+        processes and a persistent file, named with the full hash and
+        containing the spec, in a subdirectory of the database to enable
+        persistence across overlapping but separate related build processes.
+
+        The failure lock file, ``spack.store.db.prefix_failures``, lives
+        alongside the install DB. ``n`` is the sys.maxsize-bit prefix of the
+        associated DAG hash to make the likelihood of collision very low with
+        no cleanup required.
+        """
+        # Dump the spec to the failure file for (manual) debugging purposes
+        path = self._failed_spec_path(spec)
+        with open(path, 'w') as f:
+            spec.to_json(f)
+
+        # Also ensure a failure lock is taken to prevent cleanup removal
+        # of failure status information during a concurrent parallel build.
+        err = 'Unable to mark {0.name} as failed.'
+
+        prefix = spec.prefix
+        if prefix not in self._prefix_failures:
+            mark = Lock(
+                self.prefix_fail_path,
+                start=spec.dag_hash_bit_prefix(bit_length(sys.maxsize)),
+                length=1,
+                default_timeout=self.package_lock_timeout, desc=spec.name)
+
+            try:
+                mark.acquire_write()
+                tty.debug('PID {0} succeeded in marking failure for {1}'
+                          .format(os.getpid(), spec.name))
```

Comment:
Wondering if the PID should not be added directly to `tty.debug` as it is the timestamp. @tgamblin

Reply:
Good question.  Does look "messy" not having a standard.

Reply:
I'd save this for some future PR unless there is some reason we want to include it here -- there's already a lot in this one.